### ISSUE TITLE
Question: How does one scale karafka to multiple threads per consumer group?

### ISSUE THREAD

**Issue by paneq** [karafka/karafka#481]
Can it be achieved by running multiple processes of karafka?

The concurrency wiki page mentions that one consumer group will use a single thread. I assume that means a single thread for processing all partitions of a topic. Is there a way to scale to more or do we need to wait for #458 ?

**Comment by mensfeld** [comment]
## TL;DR

> Can it be achieved by running multiple processes of karafka?

Yes.

## Long answer

> The concurrency wiki page mentions that one consumer group will use a single thread.

Yes. The consumer group will run __your__ business logic within a single thread, however the data is being prefetched from a separate topic, utilizing the Ruby GIL property of not locking on IO. That means, that most of the time you should have data upfront in case of a lag.

>  I assume that means a single thread for processing all partitions of a topic.

It depends on how many Karafka processes you are running. Your statement is true only when running a single process. If you run more, the processing will distribute evenly across all the processes.

Let's take following scenario:

- Single topic
- 6 partitions

When you run a single Karafka process with a single consumer group it will consume all the partitions.

However if you run 2 processes __each__ of them will consume __3__ partitions.

If you run 7 or more processes, beyond the number of partitions, those will be left redundant.

If this answer satisfies you, please close the issue.

**Comment by paneq** [comment]
All clear. Much appreciated, consider adding it to the "Concurrency" section on wiki.

**Comment by mensfeld** [comment]
Yup. Updated. Thanks!

**Comment by zainonrails** [comment]
@mensfeld 

> However if you run 2 processes each of them will consume 3 partitions.
It means every karafka process is actually a consumer process because Kafka redistributes partitions on the basis of consumers.
Or is it handled differently through karafka because we can't guarantee if all the partitions are divided equally among consumers?

correct me if i am wrong.

P.s And how do increase karafka process? by running mutliple karafka servers?

**Comment by mensfeld** [comment]
> It means every karafka process is actually a consumer process because Kafka redistributes partitions on the basis of consumers.

Yes and no. Every Karafka process is at least one consumer as you can define several consumer groups (or you can repeat the same consumer group several times to achieve multi threading).

> Or is it handled differently through karafka because we can't guarantee if all the partitions are divided equally among consumers?

That's on Kafkas side.

> P.s And how do increase karafka process? by running mutliple karafka servers?

Either by defining the same consumer group several times within one Ruby process or by running several processes.
